spring-cloud-gcp-autoconfigure/src/main/java/org/springframework/cloud/gcp/autoconfigure/sql/AppEngineJdbcInfoProviderAutoConfiguration.java

```diff
@@ -0,0 +1,62 @@
+/*
+ *  Copyright 2018 original author or authors.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.springframework.cloud.gcp.autoconfigure.sql;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+import org.springframework.boot.autoconfigure.AutoConfigureBefore;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.cloud.gcp.autoconfigure.core.AppEngineCondition;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Conditional;
+import org.springframework.context.annotation.Configuration;
+
+/**
+ * Provides the {@link CloudSqlJdbcInfoProvider} for apps running on Google App Engine.
+ *
+ * @author João André Martins
+ */
+@Configuration
+@ConditionalOnClass({com.google.cloud.sql.mysql.SocketFactory.class})
+@ConditionalOnProperty(
+		name = "spring.cloud.gcp.sql.enabled", havingValue = "true", matchIfMissing = true)
+@AutoConfigureBefore(GcpCloudSqlAutoConfiguration.class)
+public class AppEngineJdbcInfoProviderAutoConfiguration {
+
+	private static final Log LOGGER =
+			LogFactory.getLog(AppEngineJdbcInfoProviderAutoConfiguration.class);
+
+	@Bean
+	@ConditionalOnMissingBean(CloudSqlJdbcInfoProvider.class)
+	@Conditional(AppEngineCondition.class)
+	public CloudSqlJdbcInfoProvider appengineCloudSqlJdbcInfoProvider(
+			GcpCloudSqlProperties gcpCloudSqlProperties) {
+		CloudSqlJdbcInfoProvider appEngineProvider =
+				new AppEngineCloudSqlJdbcInfoProvider(gcpCloudSqlProperties);
+
+		if (LOGGER.isInfoEnabled()) {
+			LOGGER.info("App Engine JdbcUrl provider. Connecting to "
+					+ appEngineProvider.getJdbcUrl() + " with driver "
```

Comment:
It probably has nothing to do with this PR per se but I find this a bit confusing. The only condition is on the enabled property (and classpath check).

And then, this claim that AppEngine is "configured". But if the user did not, you'll get `null` for those, isn't it? Or is this filled in by GAE?

Reply:
I don't think I understand what you mean.

The condition is simply for logging. The user actually doesn't have to configure almost anything, we already provide the JDBC URL and driver class name so they don't have to.

Reply:
I think we need to take a step back. Is there a description of this feature somewhere?

Reply:
We have some explanation of the feature in the [README](https://github.com/spring-cloud/spring-cloud-gcp/tree/master/spring-cloud-gcp-starters/spring-cloud-gcp-starter-sql). I'm not sure if it's sufficient though.

Reply:
This feature has been working for us pretty well, so far.
I'm not sure if anyone is using it, but we haven't heard any issues from users either.
The [refdoc](https://docs.spring.io/spring-cloud-gcp/docs/1.0.0.M1/reference/htmlsingle/#_spring_jdbc) also covers the feature in some detail.

I suggest if there are any issues with the way the feature works, we file it as a separate issue.
This PR is intended to split the starters into two, not fix retroactive issues.

Reply:
> This PR is intended to split the starters into two, not fix retroactive issues.

OK, I'll stop commenting  here. The problem is that this PR is so large that it is very hard to track what is off topic. I'll review the code once the structure has been stabilized then.